src/analyses/goto_check.cpp

```diff
@@ -1851,30 +1849,54 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
     return {};
 }
 
-/// Set a Boolean flag to a new value (via `set_flag`) and restore the previous
-/// value when the entire object goes out of scope.
+/// \brief Set a Boolean flag to a new value (via `set_flag`) and restore 
+/// the previous value when the entire object goes out of scope.
+///
+/// \remarks Calls to set_value are tracked to allow detecting doubles sets
+/// with different values and trigger an INVARIANT.
 class flag_resett
 {
 public:
-  /// Store the current value of \p flag and then set its value to \p new_value.
-  void set_flag(bool &flag, bool new_value)
+  /// \brief Store the current value of \p flag and
+  /// then set its value to \p new_value.
+  ///
+  /// \remarks an INVARIANT triggers iff the flag is set
+  /// more than once with different values.
+  void set_flag(
+    bool &flag,
+    bool new_value,
+    const irep_idt &flag_name,
+    const goto_programt::instructiont &instruction)
   {
+    INVARIANT(
+      !(seen(&flag) && flag != new_value),
+      "Flag " + id2string(flag_name) +
+        " set twice with incompatible values "
+        " at \n" +
+        instruction.source_location().pretty());
     if(flag != new_value)
     {
-      flags_to_reset.emplace_back(&flag, flag);
+      flags_to_reset.insert(std::pair<bool *, bool>(&flag, flag));
       flag = new_value;
     }
+    return;
   }
 
-  /// Restore the values of all flags that have been modified via `set_flag`.
+  /// \brief Restore the values of all flags that have been
+  /// modified via `set_flag`.
   ~flag_resett()
   {
     for(const auto &flag_pair : flags_to_reset)
       *flag_pair.first = flag_pair.second;
   }
 
 private:
-  std::list<std::pair<bool *, bool>> flags_to_reset;
+  /// \brief True iff flag has already been seen by \ref set_flag
+  bool seen(bool *flag)
```

Comment:
Not that it matters a whole lot, but the parameter type should be `const bool *` to make clear that no modification will occur. Also, the method should be marked `const`.

Reply:
method removed